File: get-weather.pl

:- dynamic latitudine/1, longitudine/1, url/1.

inizializza_lat_lon(Latitudine, Longitudine) :-
    assertz(latitudine(Latitudine)),
    assertz(longitudine(Longitudine)).

crea_url :-
    latitudine(Lat),
    longitudine(Lon),
    atom_concat('https://api.openweathermap.org/data/2.5/weather?lat=', Lat, Url1),
    atom_concat(Url1,'&lon=',Url2),
    atom_concat(Url2,Lon,Url3),
    atom_concat(Url3,'&appid=9525a467ef22974bc25346d4bc02de69',Url),
    assertz(url(Url)).

:- use_module(library(http/http_client)).

calcola_meteo() :-
    url(Url),
    http_get(Url, Response, []),
    atom_codes(Data, Response),
    write(Data).

%%test

%%1
%| ?- inizializza_lat_lon('45.46', '9.19').
%yes

%| ?- crea_url.
%yes

%| ?- calcola_meteo.
%{ "coord": { "lon": 9.19, "lat": 45.46 }, "weather": [ { "id": 802, "main": "Clouds", "description": "scattered clouds", "icon": "03d" } ], "base": "stations", "main": { "temp": 290.95, "feels_like": 290.6, "temp_min": 289.45, "temp_max": 292.46, "pressure": 1020, "humidity": 66 }, "visibility": 10000, "wind": { "speed": 1.34, "deg": 160 }, "clouds": { "all": 44 }, "dt": 1651727253, "sys": { "type": 2, "id": 2009385, "country": "IT", "sunrise": 1651695405, "sunset": 1651748492 }, "timezone": 7200


%%2

%?- inizializza_lat_lon('45.4642', '9.1900').
%true.

%?- crea_url.
%true.

%?- calcola_meteo.
%{ "coord":{ "lon":9.19, "lat":45.4642 }, "weather":[ { "id":800, "main":"Clear", "description":"clear sky", "icon":"01d" } ], "base":"stations", "main":{ "temp":292.87, "feels_like":291.52, "temp_min":291.15, "temp_max":294.59, "pressure":1012, "humidity":43 }, "visibility":10000, "wind":{ "speed":2.06, "deg":120 }, "clouds":{ "all":0 }, "dt":1620814447, "sys":{ "type":2, "id":2014603, "country":"IT", "sunrise":1620786364, "sunset":1620842344 }, "timezone":7200, "id":3169070, "name":"Milan", "cod":200 }
%true.

File: haskell/get-weather-.hs
import Control.Monad.IO.Class
import Data.Aeson
import Network.HTTP.Req


main :: IO ()
-- You can either make your monad an instance of 'MonadHttp', or use
-- 'runReq' in any IO-enabled monad without defining new instances.
main = runReq defaultHttpConfig $ do
data AppState = AppState
  { latitude :: Maybe String
  , longitude :: Maybe String
  , url :: Maybe String
  }

initializeLatLon :: State AppState ()
initializeLatLon = do
  put (AppState { latitude = Nothing, longitude = Nothing, url = Nothing })
  lat <- liftIO $ putStr "Enter latitude: " >> getLine
  lon <- liftIO $ putStr "Enter longitude: " >> getLine
  modify (\s -> s { latitude = Just lat, longitude = Just lon })

createUrl :: State AppState ()
createUrl = do
  lat <- gets latitude
  lon <- gets longitude
  case (lat, lon) of
    (Just lat', Just lon') ->
      modify (\s -> s { url = Just $ "https://api.openweathermap.org/data/2.5/weather?lat=" ++ lat' ++ "&lon=" ++ lon' ++ "&appid=9525a467ef22974bc25346d4bc02de69" })
    _ -> return ()

getUrl :: AppState -> Maybe String
getUrl = url

getWeatherData :: IO ()
getWeatherData = do
  let initialState = AppState { latitude = Nothing, longitude = Nothing, url = Nothing }
      finalState = execState (initializeLatLon >> createUrl) initialState
  case getUrl finalState of
    Just url' -> do
      manager <- newTlsManager
      request <- parseRequest url'
      response <- httpLbs request manager
      putStrLn $ responseBody response
    Nothing -> putStrLn "Error: Missing latitude/longitude"


--test
-- $ runhaskell get-weather.hs

--1
--Enter latitude: 37.7749
--Enter longitude: -122.4194
--{"coord":{"lon":-122.42,"lat":37.77},"weather":[{"id":800,"main":"Clear","description":"clear sky","icon":"01d"}],"base":"stations","main":{"temp":285.22,"feels_like":284.42,"temp_min":283.95,"temp_max":286.15,"pressure":1014,"humidity":65},"visibility":10000,"wind":{"speed":3.6,"deg":290},"clouds":{"all":1},"dt":1555501564,"sys":{"type":1,"id":5122,"message":0.0139,"country":"US","sunrise":1555471623,"sunset":1555520683},"timezone":-25200,"id":420006353,"name":"San Francisco","cod":200}
